### ISSUE TITLE
Weird slow backward pass of gpu

### ISSUE THREAD

**Issue by Jingfeng0705** [x-cls/superclass#11]
Happy New Year to the authors! I encountered an issue during training. The configuration I used for training was 8 H800 GPUs in a Slurm cluster, a batch size of 16384, and 6 dataloader workers. I found that in the early stages of training, the main bottleneck was dataloading (which I suspect the streaming data loader needs time to warm up). However, as the dataloading speed increases, the bottleneck shifts to the backward pass on GPU, especially GPU0. Every six iterations form a group, and only the backward pass in the first iteration of each group takes a significantly long time, while the remaining five iterations are very fast. The size of each group depends on the number of workers in the data loader. Have the authors encountered a similar situation? 

I have also experienced the same problem when testing on a non-Slurm setup with eight 3090 GPUs. My initial suspect was the communication latency between GPUs, but I checked NCCL_INFO and found that the data transfer time was negligible. 

The First few iterations⬇️
![Screenshot 2024-12-31 at 12 13 07](https://github.com/user-attachments/assets/50ff7eca-8699-476e-8898-708c90dbeed6)
Later iterations⬇️
![Screenshot 2024-12-31 at 12 13 20](https://github.com/user-attachments/assets/03c70871-4775-4dbc-99a1-0aba8db935f9)

Thanks for your time!!

**Comment by speedinghzl** [comment]
@Jingfeng0705 Happy new year! 
I didn't encounter this issue. I conducted the experiments on V100 and A100. The speed of training is stable and It won't fluctuate in speed. Have you tried training a CLIP model of the same size to see if it encounters the same issue?"


There are some logs of training CLIP-B/16 with 8 V100 for reference.

> 2024-11-28,04:39:18 | INFO | Train Epoch: 0 [400564224/512000000 (78%)] Data (t): 12.822 Batch (t): 14.049, 1781.07/s, 222.633/s/gpu LR: 0.000057 Logit Scale: 81.228 Contrastive_loss: 0.66678 (1.6251) Loss: 0.66678 (1.6251)
2024-11-28,04:40:24 | WARNING | Handling webdataset error (OSError('image file is truncated (12 bytes not processed)')). Ignoring.
2024-11-28,04:46:47 | WARNING | Handling webdataset error (OSError('image file is truncated (22 bytes not processed)')). Ignoring.
2024-11-28,04:49:13 | INFO | Train Epoch: 0 [401612800/512000000 (78%)] Data (t): 1.091 Batch (t): 4.644, 1802.44/s, 225.305/s/gpu LR: 0.000056 Logit Scale: 81.293 Contrastive_loss: 0.51701 (1.6222) Loss: 0.51701 (1.6222)
2024-11-28,04:58:49 | INFO | Train Epoch: 0 [402661376/512000000 (79%)] Data (t): 0.751 Batch (t): 4.507, 1846.45/s, 230.807/s/gpu LR: 0.000055 Logit Scale: 81.326 Contrastive_loss: 0.59680 (1.6195) Loss: 0.59680 (1.6195)
2024-11-28,05:01:25 | WARNING | Handling webdataset error (OSError('image file is truncated (107 bytes not processed)')). Ignoring.
2024-11-28,05:01:31 | WARNING | Handling webdataset error (OSError('image file is truncated (77 bytes not processed)')). Ignoring.
2024-11-28,05:04:41 | WARNING | Handling webdataset error (OSError('image file is truncated (28 bytes not processed)')). Ignoring.
2024-11-28,05:08:35 | INFO | Train Epoch: 0 [403709952/512000000 (79%)] Data (t): 1.615 Batch (t): 4.576, 1838.25/s, 229.781/s/gpu LR: 0.000054 Logit Scale: 81.380 Contrastive_loss: 0.64415 (1.6170) Loss: 0.64415 (1.6170)
2024-11-28,05:18:12 | INFO | Train Epoch: 0 [404758528/512000000 (79%)] Data (t): 2.529 Batch (t): 4.510, 1768.67/s, 221.084/s/gpu LR: 0.000053 Logit Scale: 81.428 Contrastive_loss: 0.64633 (1.6145) Loss: 0.64633 (1.6145)
2024-11-28,05:23:58 | WARNING | Handling webdataset error (OSError('image file is truncated (33 bytes not processed)')). Ignoring.
2024-11-28,05:27:53 | INFO | Train Epoch: 0 [405807104/512000000 (79%)] Data (t): 2.992 Batch (t): 4.535, 1751.24/s, 218.905/s/gpu LR: 0.000052 Logit Scale: 81.474 Contrastive_loss: 0.70578 (1.6122) Loss: 0.70578 (1.6122)
2024-11-28,05:37:32 | INFO | Train Epoch: 0 [406855680/512000000 (79%)] Data (t): 3.294 Batch (t): 4.523, 1854.33/s, 231.791/s/gpu LR: 0.000051 Logit Scale: 81.542 Contrastive_loss: 0.64518 (1.6097) Loss: 0.64518 (1.6097)
2024-11-28,05:40:01 | WARNING | Handling webdataset error (OSError('image file is truncated (13 bytes not processed)')). Ignoring.
2024-11-28,05:47:16 | INFO | Train Epoch: 0 [407904256/512000000 (80%)] Data (t): 2.237 Batch (t): 4.566, 1819.37/s, 227.421/s/gpu LR: 0.000050 Logit Scale: 81.596 Contrastive_loss: 0.62976 (1.6072) Loss: 0.62976 (1.6072)
2024-11-28,05:56:57 | INFO | Train Epoch: 0 [408952832/512000000 (80%)] Data (t): 3.244 Batch (t): 4.538, 1828.74/s, 228.593/s/gpu LR: 0.000049 Logit Scale: 81.646 Contrastive_loss: 0.60198 (1.6046) Loss: 0.60198 (1.6046)
2024-11-28,05:57:54 | WARNING | Handling webdataset error (OSError('image file is truncated (150 bytes not processed)')). Ignoring.
2024-11-28,06:06:37 | INFO | Train Epoch: 0 [410001408/512000000 (80%)] Data (t): 3.298 Batch (t): 4.527, 1797.04/s, 224.631/s/gpu LR: 0.000048 Logit Scale: 81.704 Contrastive_loss: 0.57269 (1.6020) Loss: 0.57269 (1.6020)
2024-11-28,06:16:16 | INFO | Train Epoch: 0 [411049984/512000000 (80%)] Data (t): 1.497 Batch (t): 4.522, 1810.27/s, 226.284/s/gpu LR: 0.000047 Logit Scale: 81.761 Contrastive_loss: 0.53103 (1.5992) Loss: 0.53103 (1.5992)
2024-11-28,06:25:59 | INFO | Train Epoch: 0 [412098560/512000000 (80%)] Data (t): 1.073 Batch (t): 4.557, 1779.85/s, 222.481/s/gpu LR: 0.000046 Logit Scale: 81.814 Contrastive_loss: 0.59055 (1.5967) Loss: 0.59055 (1.5967)
2024-11-28,06:29:32 | WARNING | Handling webdataset error (OSError('image file is truncated (28 bytes not processed)')). Ignoring.
2024-11-28,06:31:40 | WARNING | Handling webdataset error (OSError('image file is truncated (44 bytes not processed)')). Ignoring.
2024-11-28,06:35:43 | INFO | Train Epoch: 0 [413147136/512000000 (81%)] Data (t): 0.681 Batch (t): 4.564, 1822.51/s, 227.813/s/gpu LR: 0.000045 Logit Scale: 81.840 Contrastive_loss: 0.58690 (1.5941) Loss: 0.58690 (1.5941)
2024-11-28,06:45:25 | INFO | Train Epoch: 0 [414195712/512000000 (81%)] Data (t): 0.684 Batch (t): 4.545, 1827.14/s, 228.393/s/gpu LR: 0.000044 Logit Scale: 81.907 Contrastive_loss: 0.64290 (1.5917) Loss: 0.64290 (1.5917)
2024-11-28,06:47:44 | WARNING | Handling webdataset error (OSError('image file is truncated (68 bytes not processed)')). Ignoring.
2024-11-28,06:55:04 | INFO | Train Epoch: 0 [415244288/512000000 (81%)] Data (t): 0.689 Batch (t): 4.524, 1805.95/s, 225.743/s/gpu LR: 0.000043 Logit Scale: 81.979 Contrastive_loss: 0.62833 (1.5893) Loss: 0.62833 (1.5893)
2024-11-28,07:04:54 | INFO | Train Epoch: 0 [416292864/512000000 (81%)] Data (t): 0.688 Batch (t): 4.609, 1837.94/s, 229.742/s/gpu LR: 0.000043 Logit Scale: 82.004 Contrastive_loss: 0.54958 (1.5867) Loss: 0.54958 (1.5867)
2024-11-28,07:10:06 | WARNING | Handling webdataset error (OSError('image file is truncated (4 bytes not processed)')).

**Comment by Jingfeng0705** [comment]
Thank you for your reply! I also tried commenting out the backward pass, so that each iteration only involves the data loading and forward pass. Strangely, this resulted in the latency shifting to the data loading phase, and I noticed it fluctuating periodically.

I was wondering if you might have any insights into what could be causing this behavior. I will continue debugging on my end and will post any findings related to this here.

Thank you for your time and help!
<img width="841" alt="Screenshot 2025-01-01 at 09 56 47" src="https://github.com/user-attachments/assets/c07b7001-2ee3-419a-9274-72ee5990d93e" />

**Comment by Jingfeng0705** [comment]
Thanks for your reply and training time reference!

**Comment by speedinghzl** [comment]
@Jingfeng0705 When using the V100 or other GPUs that do not support bf16 well, we must change the training precision to amp or fp16. The default precision during training is set to amp_bfloat16. Hope it is helpful to speedup your training.
